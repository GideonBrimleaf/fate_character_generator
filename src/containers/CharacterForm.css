input {
  width: 30vw;
  margin-top: 0.5em;
  margin-bottom: 0.5em;
  text-align: center;
}

form {
  display:flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.form-submit {
  color: #fff;
  text-transform: uppercase;
  text-decoration: none;
  background: #60a3bc;
  padding: 10px;
  border-radius: 40px;
  display: inline-block;
  border: none;
  transition: all 0.4s ease 0s;
  margin: 1rem
}