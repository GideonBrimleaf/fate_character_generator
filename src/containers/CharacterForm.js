import React, { Component } from 'react'
import '../App.css'
import './CharacterForm.css'
import '../components/CharacterDetail.css'
import '../components/CharacterStatBlock.css'
import { withRouter } from 'react-router-dom'

class CharacterForm extends Component {

  constructor(props){
    super(props)
    this.state={
      id:'',
      name:'',
      aspects: {
        highConcept:'',
        trouble:'',
        relationship: '',
        aspectOne: '',
        aspectTwo: ''
      },
      approaches: {
        careful: '',
        clever: '',
        flashy: '',
        forceful: '',
        quick: '',
        sneaky: ''
      },
      stunts: [''],
      consequences: {
        mild: '',
        moderate: '',
        severe: ''
      },
      refresh : 3
    }
  }

  handleNameChange = (event) => {
    this.setState({name:event.target.value})
  }

  handleFormChange = (event, characterStatGroup) => {
    const statGroup = this.state[characterStatGroup]
    statGroup[event.target.id] = event.target.value
    this.setState({characterStatGroup : statGroup})
  }

  handleSubmit = (event) => {
    event.preventDefault()

    const id = Date.now()
    const name = this.state.name.trim()
    const highConcept = this.state.aspects.highConcept.trim()
    const trouble = this.state.aspects.trouble.trim()

    if(!name || !highConcept || !trouble){
      return
    }

    this.props.onCharacterCreated({
      id:id,
      name:name,
      aspects: {
        highConcept:highConcept,
        trouble:trouble,
        relationship: '',
        aspectOne: '',
        aspectTwo: ''
      },
      approaches: {
        careful: '',
        clever: '',
        flashy: '',
        forceful: '',
        quick: '',
        sneaky: ''
      },
      stunts: [],
      consequences: {
        mild: '',
        moderate: '',
        severe: ''
      },
      refresh : 3
    })

    this.setState({
      id:'',
      name:'',
      aspects: {
        highConcept:'',
        trouble:''
      },
      approaches: {
        careful: '',
        clever: '',
        flashy: '',
        forceful: '',
        quick: '',
        sneaky: ''
      },
      stunts: [],
      consequences: {
        mild: '',
        moderate: '',
        severe: ''
      },
      refresh : 3
    })

    this.props.history.push('/')
  }

  render() {
    return (
      <main>
        <h1>Add a New Character</h1>
        <form autoComplete="off" className="character-sheet character-sheet-text" onSubmit={this.handleSubmit}>
          <section className="character-stats character-form-section">
            <input 
              className="character-name character-sheet-item-primary character-name-input"
              type="text" 
              id="characterName" 
              placeholder="Character Name" 
              value={this.state.name}
              onChange={this.handleNameChange}
              required
            />
            <label htmlFor="refresh" className="refresh-form-header character-sheet-text">Refesh:</label>
            <input
              className="character-refresh-input"
              type="text"
              id="refresh"
            />
          </section>
          <section className="character-stats character-form-section">
            <table className="character-sheet-item-primary">
              <caption>Aspects</caption>
              <tbody>
                <tr>
                  <td className="stat-name">
                    <label htmlFor="highConcept">High Concept:</label>
                  </td>
                  <td className="form-input">
                    <input
                      className="character-sheet-text primary-input"
                      type="text" 
                      id="highConcept" 
                      placeholder="Make sure they sound banging!" 
                      value={this.state.aspects.highConcept}
                      onChange={(event) => this.handleFormChange(event, 'aspects')}
                      required
                    />
                  </td>
                </tr>
                <tr>
                  <td className="stat-name">
                    <label className="form-label" htmlFor="trouble">Trouble:</label>
                  </td>
                  <td className="form-input">
                    <input
                      className="character-sheet-text primary-input"
                      type="text" 
                      id="trouble" 
                      placeholder="Make it dark and mysterious" 
                      value={this.state.aspects.trouble}
                      onChange={(event) => this.handleFormChange(event, 'aspects')}
                      required
                    />
                  </td>
                </tr>
                <tr>
                  <td className="stat-name">
                    <label className="form-label" htmlFor="relationship">Relationship:</label>
                  </td>
                  <td className="form-input">
                    <input
                      className="character-sheet-text primary-input"
                      type="text" 
                      id="relationship" 
                      placeholder="Make sure they sound banging!" 
                      value={this.state.aspects.relationship}
                      onChange={(event) => this.handleFormChange(event, 'aspects')}
                      required
                    /> 
                  </td>
                </tr>
                <tr>
                  <td className="stat-name">
                    <label className="form-label" htmlFor="aspectOne">Aspect One:</label>
                  </td>
                  <td className="form-input">
                    <input
                      className="character-sheet-text primary-input"
                      type="text" 
                      id="aspectOne" 
                      placeholder="Make it dark and mysterious" 
                      value={this.state.aspects.aspectOne}
                      onChange={(event) => this.handleFormChange(event, 'aspects')}
                      required
                    />
                  </td>
                </tr>
                <tr>
                  <td className="stat-name">
                    <label className="form-label" htmlFor="aspectTwo">Aspect Two:</label>
                  </td>
                  <td className="form-input">
                    <input
                      className="character-sheet-text primary-input"
                      type="text" 
                      id="aspectTwo" 
                      placeholder="Make it dark and mysterious" 
                      value={this.state.aspects.aspectTwo}
                      onChange={(event) => this.handleFormChange(event, 'aspects')}
                      required
                    />
                  </td>
                </tr>
              </tbody>
            </table>
          </section>
          <input className="primary-button" type="submit" value="Add Character"/>
        </form>
      </main>
    )
  }
}

export default withRouter(CharacterForm)