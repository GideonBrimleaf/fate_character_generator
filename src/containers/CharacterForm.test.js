/* eslint-disable react/jsx-props-no-spreading */

import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, Route } from 'react-router-dom';

import { createFakeCharacter } from '../test/fakes/character.fake';
import CharacterForm from './CharacterForm';

// TODO - this.state.name bombs because state is undefined (set to the result of an array find)
it.skip('renders correctly when a character does not exist', () => {
  // Arrange
  const props = {
    characters: [],
  }

  // Act
  const result = render(
    <MemoryRouter initialEntries={['/character/1234/edit']}>
      <Route exact path="/character/:characterId/edit">
        <CharacterForm {...props} />
      </Route>
    </MemoryRouter>,
  );

  // Assert
  expect(result.asFragment()).toMatchSnapshot();
});

it('renders correctly when a character exists', () => {
  // Arrange
  const id = 1234;

  const props = {
    characters: [createFakeCharacter(id)],
  }

  // Act
  const result = render(
    <MemoryRouter initialEntries={[`/character/${id}/edit`]}>
      <Route exact path="/character/:characterId/edit">
        <CharacterForm {...props} />
      </Route>
    </MemoryRouter>,
  );

  // Assert
  expect(result.asFragment()).toMatchSnapshot();
});
