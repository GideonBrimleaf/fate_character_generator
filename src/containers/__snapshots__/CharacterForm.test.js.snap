// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when a character exists 1`] = `
<DocumentFragment>
  <main>
    <form
      autocomplete="off"
      class="character-sheet character-sheet-text"
    >
      <section
        class="character-stats"
      >
        <input
          class="character-sheet-item-primary character-name-input"
          id="name"
          placeholder="New Character"
          required=""
          type="text"
          value="1234 name"
        />
        <h5
          class="refresh-form-header character-sheet-text character-sheet-item-secondary character-sheet-header"
        >
          <label
            for="refresh"
          >
            Refresh:
            <input
              class="character-refresh-input"
              id="refresh"
              type="text"
              value="3"
            />
          </label>
        </h5>
      </section>
      <section
        class="character-stats"
      >
        <article
          class="character-sheet-item-primary"
        >
          <table>
            <caption>
              Aspects
            </caption>
            <tbody>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="highConcept"
                  >
                    High Concept:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="highConcept"
                    type="text"
                    value="1234 highConcept"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="trouble"
                  >
                    Trouble:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="trouble"
                    type="text"
                    value="1234 trouble"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="relationship"
                  >
                    Relationship:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="relationship"
                    type="text"
                    value="1234 relationship"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="aspectOne"
                  >
                    Aspect One:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="aspectOne"
                    type="text"
                    value="1234 aspectOne"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="aspectTwo"
                  >
                    Aspect Two:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="aspectTwo"
                    type="text"
                    value="1234 aspectTwo"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </article>
        <article
          class="character-sheet-item-secondary"
        >
          <table>
            <caption>
              Approaches
            </caption>
            <tbody>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="careful"
                  >
                    Careful:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text secondary-input"
                    id="careful"
                    type="text"
                    value="1234 careful"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="clever"
                  >
                    Clever:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text secondary-input"
                    id="clever"
                    type="text"
                    value="1234 clever"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="flashy"
                  >
                    Flashy:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text secondary-input"
                    id="flashy"
                    type="text"
                    value="1234 flashy"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="forceful"
                  >
                    Forceful:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text secondary-input"
                    id="forceful"
                    type="text"
                    value="1234 forceful"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="quick"
                  >
                    Quick:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text secondary-input"
                    id="quick"
                    type="text"
                    value="1234 quick"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="sneaky"
                  >
                    Sneaky:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text secondary-input"
                    id="sneaky"
                    type="text"
                    value="1234 sneaky"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </section>
      <section
        class="character-stats"
      >
        <article
          class="character-sheet-item-primary character-form-item-primary"
        >
          <h5
            class="character-sheet-header stunt-header character-sheet-text"
          >
            Stunts
          </h5>
          <ul>
            <li>
              <input
                class="stunt-input character-sheet-text"
                id="0"
                placeholder="Stunt 1"
                type="text"
                value=""
              />
            </li>
            <li>
              <input
                class="stunt-input character-sheet-text"
                id="1"
                placeholder="Stunt 2"
                type="text"
                value=""
              />
            </li>
            <li>
              <input
                class="stunt-input character-sheet-text"
                id="2"
                placeholder="Stunt 3"
                type="text"
                value=""
              />
            </li>
          </ul>
        </article>
      </section>
      <section
        class="character-stats"
      >
        <article
          class="character-sheet-item-primary"
        >
          <table>
            <caption>
              Consequences
            </caption>
            <tbody>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="mild"
                  >
                    Mild:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="mild"
                    type="text"
                    value="1234 mild"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="moderate"
                  >
                    Moderate:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="moderate"
                    type="text"
                    value="1234 moderate"
                  />
                </td>
              </tr>
              <tr>
                <td
                  class="stat-name"
                >
                  <label
                    for="severe"
                  >
                    Severe:
                  </label>
                </td>
                <td
                  class="form-input"
                >
                  <input
                    class="character-sheet-text primary-input"
                    id="severe"
                    type="text"
                    value="1234 severe"
                  />
                </td>
              </tr>
            </tbody>
          </table>
        </article>
        <article
          class="character-sheet-item-secondary"
        >
          <table>
            <caption>
              Stress
            </caption>
            <tbody>
              <tr>
                <td>
                  <button
                    class="stress-box "
                    type="button"
                  >
                    1
                  </button>
                </td>
                <td>
                  <button
                    class="stress-box "
                    type="button"
                  >
                    2
                  </button>
                </td>
                <td>
                  <button
                    class="stress-box "
                    type="button"
                  >
                    3
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </article>
      </section>
      <input
        class="primary-button"
        type="submit"
        value="Update Character"
      />
      <a
        class="primary-button"
        href="/character/1234"
      >
        Cancel
      </a>
    </form>
  </main>
</DocumentFragment>
`;
